architecting-cloud-native-dotnet-apps-for-azure/observability-patterns.md

```diff
@@ -3,12 +3,57 @@ title: Observability patterns
+
+## Logging
+The first pattern is logging. This is simply having applications write down what it is they are doing. Any time failures or problems occur in production the goal should be to reproduce the conditions under which the failures occurred in a non-production environment, typically called Dev. Having good logs in place provides a road map for developers to follow in order to duplicate problems in an environment which can be tested and experimented with.
+
+Every programming language has tooling which permits writing logs out and typically the overhead for writing these logs is quite low. Many of the logging libraries provide for logging with a variety of different criticalities which can be tuned at run time. For instance the Serilog library, a popular structured logging library for .NET, provides logging levels
+
+* Verbose
+* Debug
+* Information
+* Warning
+* Error
+* Fatal
+
+These many log levels provide granularity in logging. When the application is functioning properly in production it may be configured to only log importing messages. When the application is misbehaving then the log level can be increased so more verbose logs are gathered.  This balances performance and debuggability. 
+
+The high performance of logging tools and the tunability of verbosity should encourage developers to log as frequently as possible. Many favour a pattern of logging entry and exit from each method. This may sound like overkill but it is very infrequent that developers will wish for less logging. In fact it is not uncommon to perform deployments for the sole purpose of adding logging around a particularly problematic method. Err on the side of too much logging and not too little. 
+
+In traditional applications the log files were typically stored on the local machine. in fact, on Unix-like operating systems there is a folder structure defined, typically under `/var/log`, to hold any logs. Of course, the utility of logging to a flat file on a single machine is vastly reduced in a cloud environment. Applications producing logs may not have access to the local disk or the local disk may be highly transient as containers are shuffled around physical machines.
+
+Cloud applications developed with an eye to microservices also pose some challenges for the file system based loggers of yesteryear. A user interaction may now span multiple services which run on different machines. It would be very challenging to correlate the logs from a user or a session. 
+
+Finally, the number of users in some cloud applications is quite high. Imagine that each user generates a hundred lines of log when the login to an application. In isolation that is quite manageable but multiply that over a hundred thousand user and the volume of logs becomes challenging. 
```

Comment:
of log when the login -> of logs when they login
thousand user -> thousand users